recipes/scala/src/main/scala/com/adobe/platform/ml/TrainingDataLoader.scala

```diff
@@ -0,0 +1,67 @@
+import org.apache.spark.sql.types.TimestampType
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+
+class TrainingDataLoader extends DataLoader {
+
+  override def load(configProperties: ConfigProperties, sparkSession: SparkSession): DataFrame = {
+
+    val serviceToken: String = sparkSession.sparkContext.getConf.get("ML_FRAMEWORK_IMS_ML_TOKEN", "").toString
+    val userToken: String = sparkSession.sparkContext.getConf.get("ML_FRAMEWORK_IMS_TOKEN", "").toString
+    val orgId: String = sparkSession.sparkContext.getConf.get("ML_FRAMEWORK_IMS_ORG_ID", "").toString
+    val dataSetId: String = configProperties.get("dataSetId").getOrElse("5c05ba9c7276150000be60f2")
+    val batchId: String = configProperties.get("batchId").getOrElse("e1df7c83c030498685ce4e5f5072c179")
+    val apiKey:String = configProperties.get("apiKey").getOrElse("")
+
+    var df = sparkSession.read.format("com.adobe.platform.dataset")
+      .option(DataSetOptions.serviceToken, serviceToken)
+      .option(DataSetOptions.userToken, userToken)
+      .option(DataSetOptions.serviceApiKey, apiKey)
+      .option(DataSetOptions.orgId, orgId)
+      .option(DataSetOptions.batchId, batchId)
+      .load(dataSetId)
+
+    import sparkSession.implicits._
+
+    // Convert isHoliday to Int
+    df = df.withColumn("isHoliday", $"isHoliday".cast("Int"))
+
+    // Get the week and year from the date
+    df = df.withColumn("date", unix_timestamp($"date", "MM/dd/yy").cast(TimestampType)).withColumn("week", weekofyear($"date")).withColumn("year", year($"date"))
+
+    // Compute the weeklySalesAhead, weeklySalesLag, weeklySalesDiff
+    val window = Window.partitionBy("store").orderBy("date")
+    df = df.withColumn("weeklySalesLag", lag("weeklySales", 1).over(window)).filter(!isnull($"weeklySalesLag"))
+    df = df.withColumn("weeklySalesAhead", lag("weeklySales", -1).over(window)).filter(!isnull($"weeklySalesAhead"))
+    df = df.withColumn("weeklySalesDiff", ($"weeklySales" - $"weeklySalesLag")/$"weeklySalesLag")
+
+    // Convert the categorical data of storeType
+    val indexer = new StringIndexer().setInputCol("storeType").setOutputCol("storeTypeIndex").fit(df)
+    val indexed = indexer.transform(df)
+    val categoriesIndicies = indexed.select("storeType","storeTypeIndex").distinct.rdd
+    val categoriesMap = categoriesIndicies.map(x=>(x(0).toString,x(1).toString.toDouble)).collectAsMap
+
+    def getCategoryIndex(catMap: scala.collection.Map[String,Double], expectedValue: Double) = udf((columnValue: String) =>
+      if (catMap(columnValue) == expectedValue) 1 else 0)
+
+    df = categoriesMap.keySet.toSeq.foldLeft[DataFrame](indexed)(
+      (acc,c) =>
+        acc.withColumn(c,getCategoryIndex(categoriesMap,categoriesMap(c))($"storeType"))
+    )
+
+    df= df.drop("storeTypeIndex")
+
+    // Order by date and split the data
+    df= df.orderBy("date").withColumn("date", $"date".cast("String"))
+    df = df.filter($"date".<=("2012-01-27 00:00:00"))
```

Comment:
How will this code generalize for other data sets?  Will this binding/allowing to 2012-01-27 be problematic?

Reply:
The feature engineering done is specific to this dataset in the way that the GBT model can interpret it.